### PR TITLE
Redesign Broker's configuration setup

### PR DESCRIPTION
**Pull request by Neverlord**
This set of changes fixes some quirks about Broker's handling of configuration parameters. In particular, it removes the two environment variables `BROKER_DEBUG_VERBOSE` and `BROKER_DEBUG_LEVEL`. They had unclear semantics and did too many (undocumented) things.

The newly recommended way for users to tap into more verbose Broker output is:

- `BROKER_CONSOLE_VERBOSITY=info` for high-level events such as connections/peerings/etc
- `BROKER_CONSOLE_VERBOSITY=debug` for low-level events and seeing what get's published

Users that want the output in a file instead can use `BROKER_FILE_VERBOSITY`.

Please read the detailed commit description of fa952b8 for more technical details. This new set of variables is easy to comprehend and to explain.

Last but not least, I've added a new entry `Troubleshooting` to the documentation that walks users through the available options of getting verbose output. I left out advanced topics such as tweaking the component filter intentionally. Developers that are familiar with CAF can refer to the CAF manual in order to tap into more logging options. Broker users should not have to care about log output of CAF.

I'll also provide a followup to this PR to refine Broker's log output. In particular to make sure users get a better understanding of what went wrong in case of an error. Recent example: Broker's output gives users no clue what went wrong in case two Broker instances with incompatible versions of CAF clash (all they see is a TCP connect followed by a disconnect).

Relates #80.

### CODE REVIEW

**Comment by Neverlord** [comment]
One more note on the implementation: I'm not particularly happy with the current way of propagating errors, i.e., throwing exceptions from constructors. However, using the idiomatic CAF way (calling a function such as `parse` that returns an error) would require changes to Broker's API.

For now, I've opted to keep the API stable but I would like to see the exceptions go eventually. Currently, users can get an exception simply for declaring a variable of type `configuration` (or `endpoint`). This means these Broker types should only live inside a `try` block, which can make the API frustrating to work with and kill your application unexpectedly.

An alternative API could look like this:

```c++
int main(int argc, char** argv) {
  if (auto ep = endpoint::make(argc, argv); // passing (argc, argv) is optional
      not ep) {
    // oops, configuration error
    auto& err = ep.error();
    // ...
  } else {
    // ok, endpoint up and running
  }
}

// With explicitly using a configuration.
int main(int argc, char** argv) {
  configuration conf;
  // ... tweak default settings ...
  if (auto err = conf.init(argc, argv)) {
    // deal with error explicitly, this step is optional and otherwise performed by endoint::run
  }
  if (auto ep = endpoint::make(std::move(conf);
      not ep) {
    // ...
  } else {
    // ...
  }
}
```

In short, I would get rid of all public constructors of `endpoint`. Instead, users would have to go through `endpoint::make`. To an extent, exceptions vs. `error` / `expected` is a matter of style (though I don't think I'm alone when calling constructors that throw particularly bad style). However, a factory function returning an `expected` clearly signals to the programmer that this is an operation that can (and will) fail.

Just my 2¢. 🙂

**Comment by rsmmr** [comment]
Haven't looked at the code yet, but conceptionally this new way of configuring debug output sounds good to me. That will give users an easy way to get what's likely the most useful debug output, and once can still tweak it all further through the configuration file where helpful.

I agree that the `make` factories would be preferable over exceptions, but indeed prefer not to break the API currently. However, we could add them in addition now, and then later deprecate & remove the actors?

**Comment by Neverlord** [comment]
> add them in addition now, and then later deprecate & remove

Sounds like a plan. 👍

I'll do that in a followup.

**Comment by Neverlord** [comment]
> Minor suggestion may be to put a brief sentence in the docs aimed at Broker developers that do want to see CAF debug output.

Done. 🙂

Enabling CAF logs works exactly as you've described: compile CAF with debug/trace logs enabled and reset the filter.

**Comment by rsmmr** [comment]
Nice, merged it. 

@Neverlord A couple notes for when you clean up Broker's debug messages:

- I keep seeing errors about unexpected batch types. I saw that with the zeek-agent already, and now during my testing with `broker-pipe`:

```
# BROKER_CONSOLE_VERBOSITY=info ./broker-pipe -l 12345 -m subscribe
creating endpoint
listening on :12345 (SSL)
creating subscriber for topic(s) []
retrieving connection for 7@68FCC0C79523FF79A6159AFBF655F82238865500#1769 (SSL)
status peer_added ::ffff:127.0.0.1:45748
*** peer_added(68FCC0C79523FF79A6159AFBF655F82238865500#1769, *::ffff:127.0.0.1:45748, "handshake successful")
received unexpected batch type (dropped)
(12345)
```

Any idea what's going in there?

And one request: Can we prefix the debug output with the log level, like `[info]` and `[debug]`, or maybe even `[broker/info]` (not sure what I prefer). That way it becomes clear that it's debug output and why each line is being logged.

**Comment by Neverlord** [comment]
> Any idea what's going in there?

I'll take a look (and track it with a new issue if it's not a trivial fix).

> Can we prefix the debug output with the log level

@rsmmr yes we can! ✨

This is just a configuration option for CAF. Currently, the format string is set to "`%m`", which only prints the log message itself. Here's the full table of supported format string qualifiers:

Screen Shot 2020-01-24 at 12 50 17

Adding the level in square brackets would be simply "`[%p] %m"`". Should I add that to the config or do you think we should add more information while we're at it? I could see a timestamp being useful, for example.

**Comment by rsmmr** [comment]
Nice. Is `%c`going to say "Broker"? Then how about: `[%c/%p] %d %m`?

**Comment by Neverlord** [comment]
> Nice. Is `%c` going to say "Broker"? 

I believe we use lowercase, but yes. LGTM, I'll make a PR.
